// @stdlib/_tools/eslint/rules/no-empty-lines-between-requires/lib/main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2026 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var endsWith = require( '@stdlib/string/ends-with' );
var trim = require( '@stdlib/string/trim' );


// VARIABLES //

var REGEXP_HEADER = /^( )*[A-Z][A-Z0-9 ]+( )*(\/\/)?$/;
var rule;


// FUNCTIONS //

/**
* Checks whether a comment is a section header.
*
* @private
* @param {Object} comment - comment object
* @returns {boolean} boolean indicating if a comment denotes a section header
*/
function isSectionHeader( comment ) {
	var val;
	if (
		comment.type !== 'Line' ||
		comment.loc.start.column !== 0
	) {
		return false;
	}
	val = comment.value;
	return REGEXP_HEADER.test( val ) || endsWith( val, '//' );
}

/**
* Checks whether a VariableDeclaration contains a require call.
*
* @private
* @param {ASTNode} node - AST node
* @returns {boolean} boolean indicating if node is a require declaration
*/
function isRequireDeclaration( node ) {
	var init;
	var decl;
	if (
		node.type !== 'VariableDeclaration' ||
		!node.declarations ||
		node.declarations.length === 0
	) {
		return false;
	}
	decl = node.declarations[ 0 ];
	init = decl.init;
	if ( !init ) {
		return false;
	}
	// Handle: var foo = require( 'bar' );
	if (
		init.type === 'CallExpression' &&
		init.callee &&
		init.callee.type === 'Identifier' &&
		init.callee.name === 'require'
	) {
		return true;
	}
	// Handle: var foo = require( 'bar' ).baz;
	if (
		init.type === 'MemberExpression' &&
		init.object &&
		init.object.type === 'CallExpression' &&
		init.object.callee &&
		init.object.callee.type === 'Identifier' &&
		init.object.callee.name === 'require'
	) {
		return true;
	}
	return false;
}

/**
* Rule for validating that there are no empty lines between module-level require statements.
*
* @param {Object} context - ESLint context
* @returns {Object} validators
*/
function main( context ) {
	var source = context.getSourceCode();
	var lines = source.lines;

	/**
	* Checks whether there is a section header comment between two nodes.
	*
	* @private
	* @param {ASTNode} prevNode - previous node
	* @param {ASTNode} currNode - current node
	* @returns {boolean} boolean indicating if a section header exists between nodes
	*/
	function hasSectionHeaderBetween( prevNode, currNode ) {
		var comments;
		var i;

		comments = source.getCommentsBefore( currNode );
		for ( i = 0; i < comments.length; i++ ) {
			if (
				comments[ i ].loc.start.line > prevNode.loc.end.line &&
				isSectionHeader( comments[ i ] )
			) {
				return true;
			}
		}
		return false;
	}

	/**
	* Checks whether there are empty lines between two nodes.
	*
	* @private
	* @param {ASTNode} prevNode - previous node
	* @param {ASTNode} currNode - current node
	* @returns {boolean} boolean indicating if empty lines exist between the nodes
	*/
	function hasEmptyLinesBetween( prevNode, currNode ) {
		var startLine;
		var endLine;
		var line;
		var i;

		startLine = prevNode.loc.end.line;
		endLine = currNode.loc.start.line;

		// Check each line between the two nodes...
		for ( i = startLine; i < endLine-1; i++ ) {
			line = lines[ i ];
			if ( trim( line ) === '' ) {
				return true;
			}
		}
		return false;
	}

	/**
	* Reports the error message.
	*
	* @private
	* @param {ASTNode} prevNode - previous require node
	* @param {ASTNode} currNode - current require node
	*/
	function report( prevNode, currNode ) {
		var loc = {
			'start': {
				'line': prevNode.loc.end.line + 1,
				'column': 0
			},
			'end': {
				'line': currNode.loc.start.line,
				'column': 0
			}
		};

		context.report({
			'node': null,
			'message': 'Unexpected empty line between require statements.',
			'loc': loc,
			'fix': fix
		});

		/**
		* Fixes the lint error by removing empty lines between requires.
		*
		* @private
		* @param {Object} fixer - ESLint fixer
		* @returns {Object} fix
		*/
		function fix( fixer ) {
			var replacement;
			var startOfNext;
			var startLine;
			var endOfPrev;
			var nonEmpty;
			var endLine;
			var range;
			var line;
			var i;

			// Get the range between the two requires...
			endOfPrev = prevNode.range[ 1 ];
			startOfNext = source.getIndexFromLoc({
				'line': currNode.loc.start.line,
				'column': 0
			});

			// Collect non-empty lines between the requires...
			startLine = prevNode.loc.end.line;
			endLine = currNode.loc.start.line;
			nonEmpty = [];
			for ( i = startLine; i < endLine - 1; i++ ) {
				line = lines[ i ];
				if ( trim( line ) !== '' ) {
					nonEmpty.push( line );
				}
			}

			// Build replacement: newline + non-empty lines (each with newline) + final newline for current require
			if ( nonEmpty.length === 0 ) {
				replacement = '\n';
			} else {
				replacement = '\n' + nonEmpty.join( '\n' ) + '\n';
			}
			range = [ endOfPrev, startOfNext ];
			return fixer.replaceTextRange( range, replacement );
		}
	}

	/**
	* Validates module-level require statements.
	*
	* @private
	* @param {ASTNode} program - program node
	*/
	function validate( program ) {
		var prevRequire;
		var body;
		var node;
		var i;

		body = program.body;
		prevRequire = null;

		for ( i = 0; i < body.length; i++ ) {
			node = body[ i ];
			if ( isRequireDeclaration( node ) ) {
				if ( prevRequire !== null ) {
					// Check for empty lines between requires...
					if (
						hasEmptyLinesBetween( prevRequire, node ) &&
						!hasSectionHeaderBetween( prevRequire, node )
					) {
						report( prevRequire, node );
					}
				}
				prevRequire = node;
			}
		}
	}

	return {
		'Program:exit': validate
	};
}


// MAIN //

rule = {
	'meta': {
		'type': 'layout',
		'docs': {
			'description': 'enforce no empty lines between module-level require statements'
		},
		'fixable': 'whitespace',
		'schema': []
	},
	'create': main
};


// EXPORTS //

module.exports = rule;
